app/src/mouse/key_listener.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <drivers/behavior.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#include <zmk/event_manager.h>
+#include <zmk/events/mouse_button_state_changed.h>
+#include <zmk/events/mouse_move_state_changed.h>
+#include <zmk/events/mouse_scroll_state_changed.h>
+#include <zmk/events/mouse_tick.h>
+#include <zmk/hid.h>
+#include <zmk/endpoints.h>
+#include <zmk/mouse.h>
+
+#if !defined(CONFIG_ZMK_SPLIT) || defined(CONFIG_ZMK_SPLIT_ROLE_CENTRAL)
+static struct vector2d move_speed = {0};
+static struct vector2d scroll_speed = {0};
+static struct mouse_config move_config = (struct mouse_config){0};
+static struct mouse_config scroll_config = (struct mouse_config){0};
+static struct mouse_times start_times = (struct mouse_times){0};
+
+bool equals(const struct mouse_config *one, const struct mouse_config *other) {
+    return one->delay_ms == other->delay_ms &&
+           one->time_to_max_speed_ms == other->time_to_max_speed_ms &&
+           one->acceleration_exponent == other->acceleration_exponent;
+}
+
+static void clear_mouse_state(struct k_work *work) {
+    move_speed = (struct vector2d){0};
+    scroll_speed = (struct vector2d){0};
+    start_times = (struct mouse_times){0};
+    zmk_hid_mouse_movement_set(0, 0);
+    zmk_hid_mouse_scroll_set(0, 0);
+    LOG_DBG("Clearing state");
+}
+
+K_WORK_DEFINE(mouse_clear, &clear_mouse_state);
+
+void mouse_clear_cb(struct k_timer *dummy) {
+    k_work_submit_to_queue(zmk_mouse_work_q(), &mouse_clear);
+}
+
+static void mouse_tick_timer_handler(struct k_work *work) {
+    zmk_hid_mouse_movement_set(0, 0);
+    zmk_hid_mouse_scroll_set(0, 0);
+    LOG_DBG("Raising mouse tick event");
+    ZMK_EVENT_RAISE(
+        zmk_mouse_tick(move_speed, scroll_speed, move_config, scroll_config, start_times));
+    zmk_endpoints_send_mouse_report();
+}
+
+K_WORK_DEFINE(mouse_tick, &mouse_tick_timer_handler);
+
+void mouse_timer_cb(struct k_timer *dummy) {
+    LOG_DBG("Submitting mouse work to queue");
+    k_work_submit_to_queue(zmk_mouse_work_q(), &mouse_tick);
+}
+
+K_TIMER_DEFINE(mouse_timer, mouse_timer_cb, mouse_clear_cb);
```

Comment:
The part that initially made me bang my head against the wall with all this is the fact that Zephyr sometimes seems to de-prioritize these timers if the system is under load.
This is (as far as I was able to understand) the primary cause of the instability with mouse movement - reports can be sent out at random intervals or dropped entirely.
I was exploring re-writing this system to use a dedicated thread, but I wasn't skilled enough at C at the time.
I'm not sure this basis for the tick system is robust enough for general use.

Reply:
So... timers should be using the hardware timers to trigger, and happen in ISR context. The *queue* has a certain priority, can be preempted, etc.  Tweaking that thread priority might help, but I'll play a bit to see what I can determine. Thanks for the insight!

Reply:
This is most certainly not definitive - it's a half-remembered conclusion from an investigation I wasn't skilled enough to perform two years ago, just food for thought. I distinctly remember reports being delayed or dropped and that causing instability. `hog.c:335` might also be worth taking a look at.